messaging/src/main/java/org/axonframework/eventhandling/pooled/Coordinator.java

```diff
@@ -853,7 +853,17 @@ private Map<Segment, TrackingToken> claimNewSegments() {
                                                       .filter(segment -> !workPackages.containsKey(segment.getSegmentId()))
                                                       .collect(Collectors.toList());
 
-            int maxSegmentsToClaim = maxClaimedSegments - workPackages.size();
+            int maxSegmentsPerNode = maxSegmentProvider.apply(name);
+
+            if(segments.size() > maxSegmentsPerNode) {
```

Comment:
Adding this to the `claimNewSegments()` method means it will do more than the function name implies. I would rather have that in a new function, something like `releaseSegmentsIfTooManyClaimed()`. Then add it in the loop something like this:
```java
            if (eventStream == null || unclaimedSegmentValidationThreshold <= clock.instant().toEpochMilli()) {
                // Claim new segments, construct work packages per new segment, and open stream based on lowest segment
                unclaimedSegmentValidationThreshold = clock.instant().toEpochMilli() + tokenClaimInterval;
                try {
                    TrackingToken streamStartPosition = lastScheduledToken;
                    if (!releaseSegmentsIfTooManyClaimed()){
                        logger.debug("Processor [{}] will try to claim new segments.", name);
                        Map<Segment, TrackingToken> newSegments = claimNewSegments();

                        for (Map.Entry<Segment, TrackingToken> entry : newSegments.entrySet()) {
                            Segment segment = entry.getKey();
                            TrackingToken token = entry.getValue();
                            TrackingToken otherUnwrapped = WrappedToken.unwrapLowerBound(token);

                            streamStartPosition = streamStartPosition == null || otherUnwrapped == null
                                    ? null : streamStartPosition.lowerBound(otherUnwrapped);
                            workPackages.computeIfAbsent(segment.getSegmentId(),
                                                         wp -> workPackageFactory.apply(segment, token));
                        }

                        if (logger.isInfoEnabled() && !newSegments.isEmpty()) {
                            logger.info("Processor [{}] claimed {} new segments for processing", name, newSegments.size());
                        }
                    }
                    ensureOpenStream(streamStartPosition);
                } catch (Exception e) {
                    logger.warn("Exception occurred while Processor [{}] started work packages"
                                        + " and opened the event stream.", name, e);
                    abortAndScheduleRetry(e);
                    return;
                }
            }
```

Reply:
@gklijs  I have refactored the code as suggested. Also, all test cases are working fine. Please suggest the next step.